java/ingest/ingest-core/src/main/java/sleeper/ingest/task/IngestTask.java

```diff
@@ -13,91 +13,101 @@
+    public void run() throws InterruptedException, IOException {
+        Instant startTime = timeSupplier.get();
+        IngestTaskStatus.Builder taskStatusBuilder = IngestTaskStatus.builder().taskId(taskId).startTime(startTime);
+        LOGGER.info("Starting task {}", taskId);
         taskStatusStore.taskStarted(taskStatusBuilder.build());
-        LOGGER.info("IngestTask started at = {}", startTaskTime);
+        IngestTaskFinishedStatus.Builder taskFinishedBuilder = IngestTaskFinishedStatus.builder();
+        Instant finishTime = handleMessages(startTime, taskFinishedBuilder::addJobSummary);
+        LOGGER.info("Total number of messages processed = {}", totalNumberOfMessagesProcessed);
+        LOGGER.info("Total run time = {}", LoggedDuration.withFullOutput(startTime, finishTime));
+
+        IngestTaskStatus taskFinished = taskStatusBuilder.finished(finishTime, taskFinishedBuilder).build();
+        taskStatusStore.taskFinished(taskFinished);
+    }
 
-        IngestTaskFinishedStatus.Builder taskFinishedStatusBuilder = IngestTaskFinishedStatus.builder();
-        try {
-            jobSource.consumeJobs(job -> runJob(job, taskFinishedStatusBuilder));
-        } finally {
-            Instant finishTaskTime = getTimeNow.get();
-            taskStatusBuilder.finished(finishTaskTime, taskFinishedStatusBuilder);
-            taskStatusStore.taskFinished(taskStatusBuilder.build());
-            LOGGER.info("IngestTask finished at = {}", finishTaskTime);
+    public Instant handleMessages(Instant startTime, Consumer<RecordsProcessedSummary> summaryConsumer) throws InterruptedException, IOException {
+        while (true) {
+            Optional<MessageHandle> messageOpt = messageReceiver.receiveMessage();
+            if (!messageOpt.isPresent()) {
+                LOGGER.info("Terminating ingest task as no messages were received");
+                return timeSupplier.get();
+            }
+            try (MessageHandle message = messageOpt.get()) {
+                IngestJob job = message.getJob();
+                LOGGER.info("IngestJob is: {}", job);
+                try {
+                    Instant jobStartTime = timeSupplier.get();
+                    jobStatusStore.jobStarted(ingestJobStarted(taskId, job, jobStartTime));
+                    IngestResult result = ingester.ingest(job);
+                    Instant jobFinishTime = timeSupplier.get();
+                    RecordsProcessedSummary summary = new RecordsProcessedSummary(result.asRecordsProcessed(), jobStartTime, jobFinishTime);
+                    jobStatusStore.jobFinished(ingestJobFinished(taskId, job, summary));
+                    summaryConsumer.accept(summary);
+                    message.completed(summary);
+                    totalNumberOfMessagesProcessed++;
+                } catch (Exception e) {
+                    LOGGER.error("Failed processing ingest job, putting job back on queue", e);
+                    message.failed();
```

Comment:
Previously if an exception was thrown during an ingest task, the task would fail completely and terminate.